posts/2021/01/pytorch-tensoriterator-internals-update.md

```diff
@@ -0,0 +1,264 @@
+_must_ run in serial, then use the `serial_for_each` loop.
+
+``` cpp
+at::TensorIteratorConfig iter_config;
+iter_config
+  .add_output(out)
+  .add_input(a)
+
+  // call if output was already allocated
+  .resize_outputs(false)
+
+  // call if inputs/outputs have different types
+  .check_all_same_dtype(false);
+
+auto iter = iter_config.build();
+
+// Copies data from input into output
+auto copy_loop = [](char** data, const int64_t* strides, int64_t n) {
+  auto* out_data = data[0];
+  auto* in_data = data[1];
+
+  for (int i = 0; i < n; i++) {
+    // assume float data type for this example
+    *reinterpret_cast<float*>(out_data) += *reinterpret_cast<float*>(in_data);
+    out_data += strides[0];
+    in_data += strides[1];
+  }
+};
+
+iter.for_each(copy_loop);
+```
+
+### Using kernels for iterations
+
+Frequently we want to create a kernel that applies a simple point-wise function
+onto entire tensors. `TensorIterator` provides various such generic kernels
+that can be used for iterating over the elements of a tensor without having to
+worry about the stride, data type of the operands or details of the
+parallelism.
+
+For example, say we want to build a function that performs the point-wise
+addition of two tensors and stores the result in a third tensor. We can use the
+`cpu_kernel` function. Note that in this example we assume a tensor of `float`,
+but you can use one of the `AT_DISPATCH_ALL_TYPES*` macros to support multiple
+data types.
+
+``` cpp
+at::TensorIteratorConfig iter_config;
+iter_config
+  .add_output(c)
+  .add_input(a)
+  .add_input(b);
+
+auto iter = iter_config.build();
+
+// Element-wise add
+at::native::cpu_kernel(iter, [] (float a, float b) -> float {
+  return a + b;
+});
+```
+
+Writing the kernel in this way ensures that the value returned by the lambda
+passed to `cpu_kernel` will populate the corresponding position in the target
+output tensor.
+
+### Setting tensor iteration dimensions
+
+The value of the sizes and strides will determine which dimension of the tensor
+you will iterate over. `TensorIterator` performs optimizations to make sure
+that at least most of the iterations happen on contiguous data to take
+advantage of hierarchical cache-based memory architectures (think dimension
+coalescing and reordering for maximum data locality).
+
+A multi-dimensional tensor has a stride value for each dimension. So the stride
+that `TensorIterator` needs to use will be different depending on which
+dimension you want to iterate over. `TensorIterator` directly computes the
+strides that get passed into the loop by itself within the `build()` function.
+How exactly it computes the dimension to iterate over is something that should
+be properly understood in order to use `TensorIterator` effectively.
+
+When performing a reduction operation (see the `sum_out` code in
+[ReduceOps.cpp](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/ReduceOps.cpp#L433)),
```

Comment:
This line has changed (and probably those in the other links as well). In this case, the new correct line is 517, but perhaps you want to want to point to `make_reduction` in `ReduceOpsUtil.h`, as it looks like they refactored `sum_out`.

Reply:
`blob/master` should always be `blob/<commit-hash>`